File: app/src/main/java/com/task/paginationlist/presentation/piclist/presenter/PicListPresenter.java
package com.task.paginationlist.presentation.piclist.presenter;

import android.util.Log;

import com.arellomobile.mvp.InjectViewState;
import com.jakewharton.retrofit2.adapter.rxjava2.HttpException;
import com.task.paginationlist.data.db.models.WallpaperDb;
import com.task.paginationlist.domain.interactors.WallpaperListInteractor;
import com.task.paginationlist.presentation.piclist.views.activities.interfaces.IPicListView;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.ObservableTransformer;

@InjectViewState
public class PicListPresenter extends BasePresenter<IPicListView> {

    public static final String TAG = "PicListPresenter";

    ObservableTransformer<List<WallpaperDb>, List<WallpaperDb>> transformer;

    WallpaperListInteractor interactor;

    public PicListPresenter(WallpaperListInteractor interactor, ObservableTransformer<List<WallpaperDb>, List<WallpaperDb>> transformer) {
        this.transformer = transformer;
        this.interactor = interactor;
    }

    @Override
    protected void onFirstViewAttach() {
        addToCompositeSubscription(interactor.getFirstWallpapers()
                .compose(transformer)
                .subscribe(response -> {
                    getViewState().invalidateList(response);
                }, this::handleError));
    }

    private void handleError(Throwable throwable) {
        String commonError = "Something wrong";
        if (throwable instanceof HttpException) {
            HttpException httpEr = (HttpException) throwable;
            commonError = "Error code: " + httpEr.code() + " description: " + httpEr.getMessage();
        }
        Log.e(TAG, throwable.getMessage());
        getViewState().showError(commonError);

    }

    public void loadMoreData(int page) {
        Log.e(TAG, "LOAD MORE DATA");
        Observable<List<WallpaperDb>> obs = page == 1 ? interactor.refreshList() : interactor.loadMore(page);
        addToCompositeSubscription(obs
                .compose(transformer)
                .subscribe(response -> getViewState().invalidateList(response), this::handleError)
        );
    }

    public void restoreListState(int pos) {
        interactor.getCachedWallpapers().compose(transformer).subscribe(list -> {
            getViewState().restoreListView(list, pos);
        }, this::handleError);
    }

    public void handleOnPostEvent(int count) {
        getViewState().updateFullSizeCount(count);
    }
}

File: app/src/main/java/com/task/paginationlist/presentation/piclist/presenter/PicListPresenter.kt
package com.task.paginationlist.presentation.piclist.presenter

import android.util.Log
import com.arellomobile.mvp.InjectViewState
import com.jakewharton.retrofit2.adapter.rxjava2.HttpException
import com.task.paginationlist.data.db.models.WallpaperDb
import com.task.paginationlist.domain.interactors.WallpaperListInteractor
import com.task.paginationlist.presentation.piclist.views.activities.interfaces.IPicListView
import io.reactivex.ObservableTransformer

@InjectViewState
class PicListPresenter(
        var interactor: WallpaperListInteractor,
        var transformer: ObservableTransformer<MutableList<WallpaperDb>, MutableList<WallpaperDb>>
) : BasePresenter<IPicListView>() {

    override fun onFirstViewAttach() {
        addToCompositeSubscription(interactor.firstWallpapers
                .compose(transformer)
                .subscribe({ response -> viewState.invalidateList(response) }, { this.handleError(it) }))
    }

    private fun handleError(throwable: Throwable) {
        var commonError = "Something wrong"
        if (throwable is HttpException) {
            commonError = "Error code: " + throwable.code() + " description: " + throwable.message
        }
        Log.e(TAG, throwable.message)
        viewState.showError(commonError)

    }

    fun loadMoreData(page: Int) {
        Log.e(TAG, "LOAD MORE DATA")
        val obs = if (page == 1) interactor.refreshList() else interactor.loadMore(page)
        addToCompositeSubscription(obs
                .compose(transformer)
                .subscribe({ response -> viewState.invalidateList(response) }, { this.handleError(it) })
        )
    }

    fun restoreListState(pos: Int) {
        interactor.cachedWallpapers.compose(transformer).subscribe({ list -> viewState.restoreListView(list, pos) }, { this.handleError(it) })
    }

    fun handleOnPostEvent(count: Int) {
        viewState.updateFullSizeCount(count)
    }

    companion object {

        val TAG = "PicListPresenter"
    }
}
